base/threadingconstructs.jl

```diff
@@ -110,15 +112,67 @@ A barrier is placed at the end of the loop which waits for all tasks to finish
 execution.
 
 The `schedule` argument can be used to request a particular scheduling policy.
-The only currently supported value is `:static`, which creates one task per thread
-and divides the iterations equally among them. Specifying `:static` is an error
-if used from inside another `@threads` loop or from a thread other than 1.
+
+Except for `:static` scheduling, how the iterations are assigned to tasks, and how the tasks
+are assigned to the worker threads are undefined. The exact assignments can be different
+for each execution. The scheduling option should be considered as a hint. The loop body
+code (including any code transitively called from it) must not assume the task and worker
+thread in which they are executed. The loop body code for each iteration must be able to
+make forward progress and be free from data races independent of the state of other iterations.
+As such, synchronizations across iterations may invoke deadlock.
+
+For example, the above conditions imply that:
+
+- The lock taken in an iteration must be released within the same iteration.
+- Avoid communication between iterations using, e.g., Channels.
+- Write only to locations not shared across iterations (unless lock or atomic operation is used).
+
+Furthermore, even though `lock` and atomic operations can be useful sometimes, it is often better
+to avoid them for performance.
```

Comment:
This confuses the purpose of this docstring, which is to explain the different scheduling hints and their behaviors/constraints. Whether locks/atomics are more or less performant than some other pattern is dependent on the algorithm, access patterns, and hardware.

I would argue that this should be moved to the Performance Tips section and elaborated on in far greater detail.

Reply:
Yes, I agree that it's more of a tutorial/how-to material than reference material. But I've just seen too many examples where people slap locks in `@threads`. If the docstring explained that locks are needed for mutating shared location using a few paragraphs, it needs to counter-balance the impression by reminding the user that the lock is likely not what they want. Given that `@threads` is primarily used for case `input_length >> nthreads`, this is a reasonable assumption. So, I thought a one-sentence note like this won't hurt and practically beats purity. Ideally, we have a section in the Performance Tips that can be linked from this sentence, to expand your point that it really depends on the use case.

Reply:
Perhaps it's worth noting that the first result in the docs for `@threads` is this page, not the docstring listing. (Perhaps they should even be interleaved?) https://docs.julialang.org/en/v1/manual/multi-threading/#The-@threads-Macro

Reply:
Uhhh... the very first example really shouldn't be using `threadid` IMHO...

Reply:
This PR is now triage-approved. Do we want to remove this out to another place in the docs?

Reply:
Please feel free to remove this. I don't think tweaking the docstring should block this PR. We can discuss this separately later and add it back without breaking anything.

Reply:
```suggestion
```